python/rola/utils/gateway.py

```diff
@@ -0,0 +1,49 @@
+import requests
+from radix_engine_toolkit import PublicKeyHash
+from abc import ABC, abstractmethod
+
+from typing import List
+
+from rola.exceptions.gateway import EntityNotFound
+
+
+class MetadataProvider(ABC):
+    @abstractmethod
+    def entity_owner(self, address: str) -> List[PublicKeyHash]:
+        pass
+
+
+class GatewayMetadataProvider:
+    base_url: str
+
+    def __init__(self, base_url: str = "https://mainnet.radixdlt.com") -> None:
+        self.base_url = base_url
+
+    @classmethod
+    def for_mainnet(cls) -> "GatewayMetadataProvider":
+        return cls("https://mainnet.radixdlt.com")
+
+    @classmethod
+    def for_stokenet(cls) -> "GatewayMetadataProvider":
+        return cls("https://stokenet.radixdlt.com")
+
+    def entity_owner(self, address: str) -> PublicKeyHash.ED25519:
+        headers = {"accept": "application/json"}
+        body = {"addresses": [address]}
+        response = requests.post(
+            url=f"{self.base_url}/state/entity/details", headers=headers, json=body
+        )
+
+        if response.status_code != 200:
+            response.raise_for_status()
+
+        items = response.json().get("items", [])
+        if not items:
+            raise EntityNotFound()
+
+        metadata_items = items[0].get("metadata", {}).get("items", [])
+        owner_keys = [item for item in metadata_items if item["key"] == "owner_keys"]
+        if len(owner_keys) == 1:
+            return PublicKeyHash.ED25519(owner_keys[0]["value"]["raw_hex"])
+        else:
+            raise EntityNotFound()
```

Comment:
About this exception, I think that this would make the TypeScript implementation of Rola differ from the Python implementation.

More specifically, if no owner keys are found for the account, there is meant to be a way to verify the account by deriving the address of the account from the public key provided in the challenge and checking if it matches. This exception being the same as when an entity is not found means that its impossible to distinguish between cases when an entity was indeed not found and in cases when no owner keys exist. 

Can @xstelea confirm this?

Reply:
I can create a different exception for this case. something like OwnerKeysNotFound. what that work for you guys?

Reply:
if not mistaken, the ts version would fail if no owner keys are found

Reply:
done. code updated